src/Mvc/Mvc.Core/src/Routing/UrlHelper.cs

```diff
@@ -32,7 +32,21 @@ public UrlHelper(ActionContext actionContext)
         /// Gets the top-level <see cref="IRouter"/> associated with the current request. Generally an
         /// <see cref="IRouteCollection"/> implementation.
         /// </summary>
-        protected IRouter Router => ActionContext.RouteData.Routers[0];
+        protected IRouter Router
+        {
+            get
+            {
+                var routers = ActionContext.RouteData.Routers;
+                if (routers.Count == 0)
+                {
+                    throw new InvalidOperationException("Could not find an IRouter associated with the ActionContext. "
```

Comment:
I also tried this in the UrlHelper constructor and it broke tests. Placing the check here means that the exception will only be thrown if the IndexOutOfRangeException was already going to be thrown.

Reply:
Sounds good.